### PR TITLE
Feature/cubed sphere parallel meshgen

### LINKED ISSUE #52: Multi-PE cell-centroid mesh generator.
Followup of issue #45. Cell-centroid mesh generator that works for multiple PEs.

### LINKED ISSUE #55: Add halos to cubed sphere mesh.
Follow up number Issue #52. Add halos to the cubed sphere mesh.

### PR DESCRIPTION
**Pull request by odlomax**
This PR implements the multi-PE cubed-sphere mesh generator with halos.

The mesh generator internally treats the tiles of the cubed-sphere as separate overlapping meshes. Halos are added at tile boundaries and at partition boundaries on tile interiors.

In addition to the usual `mesh.cells()` and `mesh.nodes()` fields, the following the fields are added to mesh cells/nodes objects:

* **xy**. The xy positions of owned data follow the conventions give in the `Tiles` class. The xy projection of halo data is contiguous with the formal xy projection of the owned data. This should will make interpolation significantly simpler. Cells and nodes both have xy fields.

* **lonlat**. Cells receive lonlat coordinates as well as nodes.

* **ijt**. Cells and nodes have fields giving the (i, j, t) coordinates of the data. Halo (i, j, t) indices are extrapolated from the owned data, i.e. i and j can be less than zero and greater than N for tile-boundary halos. This information can be used to construct a structured `CubedSphere` function space, similar to `StructuredColumns`.

Utility classes were written for the mesh generator in `src/atlas/meshgenerator/detail/cubedsphere/CubedSphereUtility.h, CubedSphereUtility.cc`. The main class is `NeighbourJacobian`. This handles the xy and ij coordinate transforms between a tile and its four neighbours. I've put it here for the time being, but there's probably a better home for it. It's indispensable for finding the ij and xy positions of halos. I also believe it will be useful for the interpolation of vector fields.

_Fig. 1. The xy projection for an N12 cubed sphere mesh with `cubed_sphere` partitioner and `halo = 3` on 8 PEs. Some of the halos overlap in this projection and are not visible in the figure._
allhalo3xyz

_Fig 2. The mesh on PE 7 from Fig. 1. Note that the xy projection of the halos is contiguous with owned data projection._
part7halo3xy

_Fig. 3. As Fig. 2, but projected into xyz space using the lonlat coordinates. Note that "gaps" in the cross shaped halo distribution in Fig. 2 do not exist in 3D space._
part7halo3xyz

I've added test executable `src/tests/mesh/test_cubedsphere_meshgen.cc`

This performs two tests.
* The first test makes sure the `NeighbourJacobian` can correctly place halo points around an `N2` grid.
* The second test runs the cubed sphere mesh generator with different partitioner and halo-size configurations. It ensures that a simple test field (see below) can be written to owned nodes/cells and distributed to halo nodes cells via the Atlas halo-exchange methods.

_Fig 4. A simple test field with f(lon, lat) = sin(3 * lon) * sin(2 * lat)_
testfield

### CODE REVIEW

**Comment by odlomax** [comment]
This PR is enormous. So feel free to arrange a meeting with me if you need to discuss anything!

**Comment by odlomax** [comment]
For completeness, the same as Figs. 1, 2 and 3, but with the `equal_regions` partitioner.

ERallhalo3xyz

ERpart7halo3xy

ERpart7halo3xyz

**Comment by odlomax** [comment]
Hi @wdeconinck,

Just an update that I've added your exception handling suggestions to this `JCSDA-internal/atlas` PR. It think @MarekWlasak will need to submit `ecmwf/atlas` PRs for his periodicity class and cubed sphere partitioner before I can create an `ecmwf/atlas` PR for this.

The final mesh generator became a bit of a beast, but at least it's not as big as `StructuredMeshGenerator.cc`! 😅

File: src/atlas/mesh/actions/BuildParallelFields.cc [hunk 1]
```diff
@@ -1072,7 +1072,12 @@ Field& build_cells_remote_idx( mesh::Cells& cells, const mesh::Nodes& nodes ) {
 }
 
 void build_cells_parallel_fields( Mesh& mesh ) {
-    build_cells_remote_idx( mesh.cells(), mesh.nodes() );
+
+  bool parallel = false;
+  mesh.cells().metadata().get( "parallel", parallel );
```
Comment (MarekWlasak) [inline src/atlas/mesh/actions/BuildParallelFields.cc:1077]: Question - I presume that you are assuming here that the parallel gets only set if it exists in the cell metadata. What happens if parallel does not exist in the metadata?

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 2]
```diff
@@ -111,15 +119,6 @@ void CubedSphereMeshGenerator::generate(const Grid& grid, const grid::Distributi
...
-  // Check for correct stagger
-  const auto gridName = grid.name();
-  const auto gridStagger = gridName.substr(gridName.rfind("-") - 1, 1);
-
-  if (gridStagger != "C") {
-    throw_Exception("CubedSphereMeshGenerator can only work with a"
-    "cell-centroid grid. Try NodalCubedSphereMeshGenerator instead.");
-  }
-
   // Cast grid to cubed sphere grid.
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:122]: Why did you remove the error trap for the stagger?

Comment (MarekWlasak) [reply src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:122]: I now can see that it has been moved.

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 3]
```diff
@@ -199,53 +224,106 @@ void CubedSphereMeshGenerator::generate_mesh(const CubedSphereGrid& csGrid,
...
+    idx_t ijMax = includeHalo ? N + nHalo : N;
+
+    t = std::max(std::min(t, 5), 0);
+    j = std::max(std::min(j, ijMax), ijMin);
+    i = std::max(std::min(i, ijMax), ijMin);
+
+    const auto rowSize = (N + 2 * nHalo + 1);
+    const auto tileSize = rowSize * rowSize;
+
+    return idx2st(t * tileSize + (j + nHalo) * rowSize + i + nHalo);
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc]: spacing above is not quite right.

Comment (MarekWlasak) [reply src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc]: Scrap that - indexing is o.k. - it is part of function getNodeIdx

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 4]
```diff
@@ -199,53 +224,106 @@ void CubedSphereMeshGenerator::generate_mesh(const CubedSphereGrid& csGrid,
...
+    const auto rowSize = (N + 2 * nHalo);
+    const auto tileSize = rowSize * rowSize;
+
+    return idx2st(t * tileSize + (j + nHalo) * rowSize + i + nHalo);
+
+  };
+
+  // Return true for nodes interior to or on edge of tile.
+  const auto interiorNode = [&](idx_t j, idx_t i){
+    return i >= 0 and i < N + 1 and j >= 0 and j < N + 1;
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc]: Can you please not use alternative tokens to the standard c++ set of operators.  They are a code smell.  They were introduced because some keyboards have difficulty getting & (for instance).  You don't have this problem.  For a while I was looking for a bespoke overload. https://rules.sonarsource.com/cpp/RSPEC-3659

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 5]
```diff
@@ -199,53 +224,106 @@ void CubedSphereMeshGenerator::generate_mesh(const CubedSphereGrid& csGrid,
...
-
   // Define an index counter.
   const auto idxSum = [](const std::vector<idx_t>& idxCounts) -> idx_t {
     return std::accumulate(idxCounts.begin(), idxCounts.end(), 0);};
 
-  // Helper functions to get node and cell idx from (t, j, i).
-  const auto getNodeIdx = [&](idx_t t, idx_t j, idx_t i){
+  // Helper functions to get node idx from (t, j, i).
+  const auto getNodeIdx = [&](idx_t t, idx_t j, idx_t i, bool includeHalo){
+
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:269]: This is an example pf an auto that does not tell you what you are returning. (It is actually size_t)

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 6]
```diff
@@ -305,368 +382,611 @@ void CubedSphereMeshGenerator::generate_mesh(const CubedSphereGrid& csGrid,
...
-            cell.xy.x() + di * jac.dxByDi + dj * jac.dxByDj,
-            cell.xy.y() + di * jac.dyByDi + dj * jac.dyByDj
-          };
+          // Get owning node
+          PointIJ ijGlobal;
+          idx_t tGlobal;
 
-        // Get remote xy from tile class and local t.
-        node.remoteXy = csTiles.tileCubePeriodicity(node.localXy, t);
+          // Nodes reside on interger ij values.
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc]: Typo integer

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 7]
```diff
@@ -305,368 +382,611 @@ void CubedSphereMeshGenerator::generate_mesh(const CubedSphereGrid& csGrid,
...
+  for (idx_t t = 0; t < 6; ++t) {
 
-    // Limit range to bounds recorded earlier.
-    const auto bounds = cellBounds[idx2st(t)];
+      // Limit range to bounds recorded earlier.
+      const auto bounds = cellBounds[idx2st(t)];
+      const auto iBegin = bounds.iBegin - nHalo;
+      const auto iEnd   = bounds.iEnd   + nHalo + 1;
+      const auto jBegin = bounds.jBegin - nHalo;
+      const auto jEnd   = bounds.jEnd   + nHalo + 1;
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:642]: spacing above is a bit funny

File: src/atlas/mesh/actions/BuildHalo.cc [hunk 8]
```diff
@@ -1303,6 +1303,20 @@ void BuildHalo::operator()( int nb_elems ) {
...
+    if (mesh_.metadata().get("mesh_type", meshType)) {
+      if (meshType == "cubed_sphere") {
+        if (halo < nb_elems) {
+          const auto errMsg = "Error: cubed sphere halos must be set up at mesh"
+            "generation. Current halo size is " + std::to_string(halo) + ". "
+            "Required halo size " + std::to_string(nb_elems) + ".";
+          throw_Exception(errMsg, Here());
+        }
+      }
+    }
```
Comment (wdeconinck) [inline src/atlas/mesh/actions/BuildHalo.cc:1317]: I probably prefer a  locking mechanism instead, such as a "halo_locked" metadata rather than checking specific mesh types.

**Comment by wdeconinck** [comment]
https://github.com/ecmwf/atlas/pull/70 is merged so ready when you are  =)

**Review by MarekWlasak — APPROVED** [review]
Approval based on additional informal chat.

File: src/atlas/mesh/actions/BuildParallelFields.cc (hunk 1 above)
Comment (wdeconinck) [reply src/atlas/mesh/actions/BuildParallelFields.cc:1077]: then the value "parallel" will remain untouched, hence "parallel = false" remains.

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc (hunk 4 above)
Comment (wdeconinck) [reply src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc]: I tend to agree, though I personally like to use "not" as it is much more readable than "!" which can sometimes be overlooked.

**Comment by odlomax** [comment]
This has been superseded by https://github.com/ecmwf/atlas/pull/73 and https://github.com/JCSDA-internal/atlas/pull/64.
